[pt1][quant] Add the FP16 weight support for LSTM in dynamic_quantize

jianyuh opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26378 [WIP][pt1][quant] Add the serialization support for FP16 LSTM
* **#25975 [pt1][quant] Add the FP16 weight support for LSTM in dynamic_quantize**

We would like to add the FP16 weight support for the dynamic quantized LSTM.

Todo:
- Deduplicate the code between from_float and __init__.

Differential Revision: [D17299116](https://our.internmc.facebook.com/intern/diff/D17299116/)

jianyuh committed:
[pt1][quant] Add the FP16 weight support for LSTM in dynamic_quantize

We would like to add the FP16 weight support for the dynamic quantized LSTM

Differential Revision: [D17299116](https://our.internmc.facebook.com/intern/diff/D17299116/)

jamesr66a commented on torch/nn/quantized/dynamic/modules/rnn.py:
```diff
@@ -58,45 +59,70 @@ def __init__(self, mode, input_size, hidden_size,
+                        #
+                        #   w_ih, w_hh, b_ih, b_hh
+                        packed_weight = \
+                            torch.ops.quantized.linear_prepack(qweight)
```
This should include bias, right? 

It's probably worth it to rebase on top of my change: https://github.com/pytorch/pytorch/pull/25877

jianyuh replied:
Have rebased on top of your Diff. Right: it should include the bias.

jamesr66a commented on torch/nn/quantized/dynamic/modules/rnn.py:
```diff
@@ -203,38 +235,56 @@ def from_float(cls, mod):
-                    packed_weights.append(ret_name[0])
-                    return params, ret_name
+
+                    if dtype == torch.qint8:
```
Same comments as above

BTW we should probably deduplicate this code between from_float and the constructor

jianyuh replied:
Have rebased it. Yes: we should consider a way to deduplicate the code between `from_float` and `__init__`.

jamesr66a reviewed (commented):
Looks fine on the whole, but I think you should stack on top of my PR since it changes code you're changing pretty significantly

jianyuh committed:
Update on "[pt1][quant] Add the FP16 weight support for LSTM in dynamic_quantize"


We would like to add the FP16 weight support for the dynamic quantized LSTM

Differential Revision: [D17299116](https://our.internmc.facebook.com/intern/diff/D17299116/)

jamesr66a reviewed (changes requested):
Cool! Added some comments. In general I think we need to make sure all your fp16 code is using the new weights layout I introduced, as well as figure out that JIT issue you saw

jianyuh committed:
Update on "[pt1][quant] Add the FP16 weight support for LSTM in dynamic_quantize"


We would like to add the FP16 weight support for the dynamic quantized LSTM

Differential Revision: [D17299116](https://our.internmc.facebook.com/intern/diff/D17299116/)

jianyuh committed:
Update on "[pt1][quant] Add the FP16 weight support for LSTM in dynamic_quantize"


We would like to add the FP16 weight support for the dynamic quantized LSTM

Differential Revision: [D17299116](https://our.internmc.facebook.com/intern/diff/D17299116/)

jianyuh commented on torch/nn/quantized/dynamic/modules/rnn.py:
```diff
@@ -51,47 +52,76 @@ def __init__(self, mode, input_size, hidden_size,
+                        #
+                        #   packed_ih, packed_hh, b_ih, b_hh
+                        packed_weight = torch.fbgemm_pack_gemm_matrix_fp16(
+                            qweight.clone().float())
```
Right: I have removed that.

jianyuh committed:
Update on "[pt1][quant] Add the FP16 weight support for LSTM in dynamic_quantize"


We would like to add the FP16 weight support for the dynamic quantized LSTM

Differential Revision: [D17299116](https://our.internmc.facebook.com/intern/diff/D17299116/)

jianyuh commented on torch/nn/quantized/dynamic/modules/rnn.py:
```diff
@@ -179,23 +209,29 @@ def __setstate__(self, state):
+            raise RuntimeError('Unsupported dtype: {}'.format(dtype))
+
+        if dtype == torch.qint8:
+            # When dtype = torch.float16, we don't need weight_observer
```
Sure! Have moved it above the if statement.

jianyuh commented on torch/nn/quantized/dynamic/modules/rnn.py:
```diff
@@ -51,47 +52,76 @@ def __init__(self, mode, input_size, hidden_size,
-                    [gate_size], scale=1, zero_point=0, dtype=torch.qint32)
+                def process_weights(ihhh, layer, suffix, qweight, bias, dtype):
+                    weight_name = 'weight_{}_l{}{}'.format(ihhh, layer, suffix)
+                    bias_name = 'bias_{}_l{}{}'.format(ihhh, layer, suffix)
```
Have removed them.

jianyuh commented on torch/nn/quantized/dynamic/modules/rnn.py:
```diff
@@ -51,47 +52,76 @@ def __init__(self, mode, input_size, hidden_size,
+                    for (ih, ih_name), (hh, hh_name) in zip(zip(ih_params, ih_param_names), zip(hh_params, hh_param_names)):
+                        self.register_buffer(ih_name, torch.tensor(
+                            ih) if not isinstance(ih, torch.Tensor) else ih)
+                        self.register_buffer(hh_name, torch.tensor(
```
Updated

jianyuh commented on torch/quantization/quantize.py:
```diff
@@ -331,7 +332,11 @@ def swap_module(mod, mapping):
-            new_mod = mapping[type(mod)].from_float(mod)
+            if (dtype == torch.qint8):
+                new_mod = mapping[type(mod)].from_float(mod)
+            else:  # dtype == torch.float16
```
Updated.

jianyuh commented on test/test_quantization.py:
```diff
@@ -490,12 +490,20 @@ def test_quantized_rnn(self):
         model_int8 = quantize_dynamic(
-            model, qconfig_dynamic_dict, default_dynamic_module_mapping
+            model=model, qconfig_dict=qconfig_dynamic_dict, mapping=default_dynamic_module_mapping,
+            dtype=torch.torch.qint8
```
Fixed.

jianyuh commented on test/test_quantization.py:
```diff
@@ -538,15 +546,22 @@ def forward(self, x, hiddens):
+        for out, ref in zip(final_hiddens_fp16, ref_hid):
+            torch.testing.assert_allclose(out, ref)
+
+#         # Test save/load
```
For torch.float, we don’t need `dynamic_val`. But we need to keep the original_weight for serialization. This caused some inconsistency between qint8 and float16 serialization.

jamesr66a replied:
@jianyuh can you just make a convention to use dynamic_val to hold original_weight as a single element under fp16? Then in __setstate__ we would just branch on dtype and either pull out the dynamic vals normally (int8) or pull assert dynamic val has 1 element then just pull that out and assign to original_weight

Btw you should put the serialization testing code back

jianyuh replied:
Right: not removing the serialization testing code (the int8 serialization code is kept as it is). For the FP16 serialization, I will separate as a new PR: https://github.com/pytorch/pytorch/pull/26378 (Still WIP).

jamesr66a reviewed (approved):
I think this is good to go. Only minor comments

jianyuh committed:
Update on "[pt1][quant] Add the FP16 weight support for LSTM in dynamic_quantize"


We would like to add the FP16 weight support for the dynamic quantized LSTM.

Todo:
- Deduplicate the code between from_float and __init__.

Differential Revision: [D17299116](https://our.internmc.facebook.com/intern/diff/D17299116/)

[ghstack-poisoned]

jianyuh commented on torch/nn/quantized/dynamic/modules/rnn.py:
```diff
@@ -173,29 +194,36 @@ def __setstate__(self, state):
-            weight_observer = default_dynamic_qconfig.weight()
-        assert weight_observer.dtype == torch.qint8, 'Weight observer must have dtype torch.qint8'
+
+        if dtype != torch.qint8 and dtype != torch.float16:
```
Fixed.

jianyuh commented on torch/nn/quantized/dynamic/modules/rnn.py:
```diff
@@ -173,29 +194,36 @@ def __setstate__(self, state):
+            raise RuntimeError('Unsupported dtype: {}'.format(dtype))
+
+        # When dtype = torch.float16, we don't need weight_observer
+        if dtype == torch.qint8:
```
Fixed.

jianyuh commented on torch/quantization/quantize.py:
```diff
@@ -331,7 +332,13 @@ def swap_module(mod, mapping):
     if hasattr(mod, 'qconfig') and mod.qconfig is not None:
         if type(mod) in mapping:
-            new_mod = mapping[type(mod)].from_float(mod)
+            if dtype != torch.qint8 and dtype != torch.float16:
```
Fixed.

jianyuh replied:
Fixed. Should we only keep one copy of `supported_scalar_types` globally? Currently we need to use it in `torch/quantization/quantize.py` and `torch/nn/quantized/dynamic/modules/rnn.py`.

jianyuh commented on torch/quantization/quantize.py:
```diff
@@ -331,7 +332,13 @@ def swap_module(mod, mapping):
-            new_mod = mapping[type(mod)].from_float(mod)
+            if dtype != torch.qint8 and dtype != torch.float16:
+                 raise RuntimeError('Unsupported dtype: {}'.format(dtype))
+            if (dtype == torch.qint8):
```
Fixed.

jianyuh committed:
Update on "[pt1][quant] Add the FP16 weight support for LSTM in dynamic_quantize"


We would like to add the FP16 weight support for the dynamic quantized LSTM.

Todo:
- Deduplicate the code between from_float and __init__.

Differential Revision: [D17299116](https://our.internmc.facebook.com/intern/diff/D17299116/)

[ghstack-poisoned]
